### PR TITLE
Adding rbac tests for Cluster Action : snapshot on an RKE2 cluster

### PR DESCRIPTION
**Pull request by sowmyav27**
## Issue: 
Adding RBAC tests for snapshot creation on an RKE2 cluster
 

## Solution

- Perform snapshot action on a cluster
- Verify only Cluster owner and restricted admin have access. Cluster member, project owner and member do not have access. And get `Resource type [provisioning.cattle.io.cluster] is not updatable` error

### Manual Testing
N/A

### Automated Testing
Adding these automated tests - https://github.com/rancher/qa-tasks/issues/410

## QA Testing Considerations
N/A
 
### Regressions Considerations
N/A

### CODE REVIEW

File: tests/v2/validation/provisioning/rke2/etcd_backup_restore.go [hunk 1]
```diff
@@ -31,14 +31,14 @@ const (
...
-func createSnapshot(client *rancher.Client, clustername string, generation int, namespace string) error {
+func CreateSnapshot(client *rancher.Client, clustername string, generation int, namespace string) error {
 	clusterObj, existingSteveAPIObj, err := getProvisioningClusterByName(client, clustername, namespace)
 	if err != nil {
 		return err
 	}
 
 	clusterObj.Spec.RKEConfig.ETCDSnapshotCreate = &rkev1.ETCDSnapshotCreate{
-		Generation: generation,
+		Generation: clusterObj.Spec.RKEConfig.ETCDSnapshotCreate.Generation + generation,
```
Comment (anupama2501) [inline tests/v2/validation/provisioning/rke2/etcd_backup_restore.go:41]: lets make this + 1 and remove the integer generation from the function call as we need to update the snapshot by + 1 each time.
```suggestion
		Generation: clusterObj.Spec.RKEConfig.ETCDSnapshotCreate.Generation + 1,
```

File: tests/v2/validation/rbac/rbac_etcd_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+type ETCDRbacTestSuite struct {
+	suite.Suite
+	client             *rancher.Client
+	standardUser       *management.User
+	standardUserClient *rancher.Client
+	session            *session.Session
+	cluster            *management.Cluster
+	adminProject       *management.Project
+	ns                 string
```
Comment (anupama2501) [inline tests/v2/validation/rbac/rbac_etcd_test.go:26]: Can we name `ns` as namespace?

File: tests/v2/validation/rbac/rbac_etcd_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	err := rke2.CreateSnapshot(rb.standardUserClient, rb.clusterName, 1, rb.ns)
+
+	switch role {
+	case roleOwner, restrictedAdmin:
+		require.NoError(rb.T(), err)
+		// verify snapshot is created and is Active
+
+	case roleMember, roleProjectOwner, roleProjectMember:
+		require.Error(rb.T(), err)
+		strings.Contains(err.Error(), "is not updatable")
```
Comment (anupama2501) [inline tests/v2/validation/rbac/rbac_etcd_test.go:68]: lets add the full error like "Resource type [xxx] is not updatable", and assert the same error that we get from err.Error(). So, the code is more readable.

File: tests/v2/validation/rbac/rbac_etcd_test.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			require.NoError(rb.T(), err)
+			rb.standardUser = newUser
+			rb.T().Logf("Created user: %v", rb.standardUser.Username)
+			rb.standardUserClient, err = rb.client.AsUser(newUser)
+			require.NoError(rb.T(), err)
+
+			subSession := rb.session.NewSession()
+			defer subSession.Cleanup()
+
+			createProjectAsAdmin, err := createProject(rb.client, rb.cluster.ID)
```
Comment (anupama2501) [inline tests/v2/validation/rbac/rbac_etcd_test.go:96]: Wondering if we can remove this code line as we are not using the `createProjectAdmin` elsewhere in the validation and remove it from the `ETCDRbacTestSuite` struct as well.?

**Review by anupama2501 — CHANGES_REQUESTED** [review]
Looks good overall. have a few comments. Can we also run go fmt on the rbac_etcd_test.go, as to remove any additional spaces that were added.

File: tests/v2/validation/rbac/rbac_etcd_test.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/rancher/rancher/tests/framework/extensions/clusters"
+	"github.com/rancher/rancher/tests/framework/extensions/users"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/rancher/rancher/tests/v2/validation/provisioning/rke2"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type ETCDRbacTestSuite struct {
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_test.go:18]: can we update the test names and filename to include `snapshot` or `backup` to be consistent with other test files i.e. `rbac_etcd_backup_test.go` ... `ETCDRbacBackupTestSuite`

**Review by slickwarren — COMMENTED** [review]
overall lgtm

File: tests/v2/validation/rbac/rbac_etcd_test.go (hunk 5 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_test.go:18]: done

File: tests/v2/validation/rbac/rbac_etcd_test.go (hunk 2 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_test.go:26]: done

File: tests/v2/validation/rbac/rbac_etcd_test.go (hunk 3 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_test.go:68]: done

File: tests/v2/validation/rbac/rbac_etcd_test.go (hunk 4 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_test.go:96]: this is needed since the user needs to be added as member or owner to the project or cluster. And as that member, i perform a snapshot action on the cluster

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+			relogin, err := rb.standardUserClient.ReLogin()
+			require.NoError(rb.T(), err)
+			rb.standardUserClient = relogin
+
+		})
+
+		rb.T().Logf("Starting validations for %v", tt.role)
+
+		rb.Run("Testcase1 - Validating if user is able to take Etcd snapshot of a cluster as the role "+tt.name, func() {
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:123]: nit - can we shorten this name? i.e. `testcase1 - take etcd snapshot as `

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	require.NotEmptyf(rb.T(), rb.clusterName, "Cluster name to install should be set")
+	clusterID, err := clusters.GetClusterIDByName(rb.client, rb.clusterName)
+	require.NoError(rb.T(), err, "Error getting cluster ID")
+	rb.cluster, err = rb.client.Management.Cluster.ByID(clusterID)
+	require.NoError(rb.T(), err)
+}
+
+func (rb *ETCDRbacBackupTestSuite) ValidateEtcdSnapshotCluster(role string) {
+
+	/* Testcase1 Verify cluster members - Owner/member,  Project members - Owner/member are able to perform snapshot action on clusters
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:54]: I don't think this comment block is necessary.

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	rb.client = client
+	rb.clusterName = client.RancherConfig.ClusterName
+	require.NotEmptyf(rb.T(), rb.clusterName, "Cluster name to install should be set")
+	clusterID, err := clusters.GetClusterIDByName(rb.client, rb.clusterName)
+	require.NoError(rb.T(), err, "Error getting cluster ID")
+	rb.cluster, err = rb.client.Management.Cluster.ByID(clusterID)
+	require.NoError(rb.T(), err)
+}
+
+func (rb *ETCDRbacBackupTestSuite) ValidateEtcdSnapshotCluster(role string) {
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:52]: this function should live in `etcd_backup_restore.go` afaik, not in the `..._test.go` file. All helpers will live there instead.

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	Cluster Owner/Restricted Admin - Allowed
+	Cluster Member/Project Owner/Project Member - Not Allowed */
+	log.Infof("Creating a snapshot of the cluster as %v", role)
+
+	err := rke2.CreateSnapshot(rb.standardUserClient, rb.clusterName, 1, rb.namespace)
+
+	switch role {
+	case roleOwner, restrictedAdmin:
+		require.NoError(rb.T(), err)
+		// verify snapshot is created and is Active
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:64]: remove comment please

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+			rb.standardUser = newUser
+			rb.T().Logf("Created user: %v", rb.standardUser.Username)
+			rb.standardUserClient, err = rb.client.AsUser(newUser)
+			require.NoError(rb.T(), err)
+
+			subSession := rb.session.NewSession()
+			defer subSession.Cleanup()
+
+			createProjectAsAdmin, err := createProject(rb.client, rb.cluster.ID)
+			rb.adminProject = createProjectAsAdmin
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:97]: I think you could directly set this in the declaration above instead of using a temporary variable.

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+			require.NoError(rb.T(), err)
+
+		})
+
+		rb.Run("Adding user as "+tt.name+" to the downstream cluster.", func() {
+			//Adding created user to the downstream clusters with the specified roles.
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:103]: rm this comment please

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+				if strings.Contains(tt.role, "project") {
+					err := users.AddProjectMember(rb.client, rb.adminProject, rb.standardUser, tt.role)
+					require.NoError(rb.T(), err)
+				} else {
+					err := users.AddClusterRoleToUser(rb.client, rb.cluster, rb.standardUser, tt.role)
+					require.NoError(rb.T(), err)
+				}
+			}
+
+			relogin, err := rb.standardUserClient.ReLogin()
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:115]: you don't have to add a comment or anything here, but for my own understanding why is this necessary?

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	err := rke2.CreateSnapshot(rb.standardUserClient, rb.clusterName, 1, rb.namespace)
+
+	switch role {
+	case roleOwner, restrictedAdmin:
+		require.NoError(rb.T(), err)
+		// verify snapshot is created and is Active
+
+	case roleMember, roleProjectOwner, roleProjectMember:
+		require.Error(rb.T(), err)
+		strings.Contains(err.Error(), "Resource type [provisioning.cattle.io.cluster] is not updatable")
```
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:68]: I think this will need some validation that the strings.Contains did not error out?

**Review by slickwarren — CHANGES_REQUESTED** [review]
some refactoring requests in the new test file

**Comment by slickwarren** [comment]
also just thought of this; is there any way we can specify that this is only for rke2? Either in the filename or something similar? My understanding is, this test won't work on rke1?

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 8 above)
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:52]: Since we are performing the validations for cluster-owner/member etc here, it is preferable, to have this function in this test file.

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+				} else {
+					err := users.AddClusterRoleToUser(rb.client, rb.cluster, rb.standardUser, tt.role)
+					require.NoError(rb.T(), err)
+				}
+			}
+
+			relogin, err := rb.standardUserClient.ReLogin()
+			require.NoError(rb.T(), err)
+			rb.standardUserClient = relogin
+
```
Comment (anupama2501) [inline tests/v2/validation/rbac/rbac_etcd_backup_test.go:118]: There's an additional space here. please remove it.

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 6 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:123]: done

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 7 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:54]: removed

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 9 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:64]: done

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 11 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:103]: done

File: tests/v2/validation/provisioning/rke2/etcd_backup_restore.go [hunk 15]
```diff
@@ -32,14 +32,26 @@ const (
 	etcdnodeCount                = 3
 )
 
-func createSnapshot(client *rancher.Client, clustername string, generation int, namespace string) error {
+func CreateSnapshot(client *rancher.Client, clustername string, generation int, namespace string) error {
```
[the remark this answers is not in the document]
Comment (sowmyav27) [reply tests/v2/validation/provisioning/rke2/etcd_backup_restore.go:35]: done. I had this initially, but missed it during some changes with my PR

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 14 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go]: done

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go]: done

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 12 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go:109]: > This is necessary because when we add a user to a cluster, the roles are not refreshed on that user. In order for the user to have all the roles in the cluster that they were added to we would need to make them relogin [ which updates their schemas]

As per @anupama2501

File: tests/v2/validation/provisioning/rke2/etcd_backup_restore.go [hunk 16]
```diff
@@ -32,14 +32,26 @@ const (
 	etcdnodeCount                = 3
 )
 
-func createSnapshot(client *rancher.Client, clustername string, generation int, namespace string) error {
+func CreateSnapshot(client *rancher.Client, clustername string, namespace string) error {
```
Comment (igomez06) [inline tests/v2/validation/provisioning/rke2/etcd_backup_restore.go:35]: Instead of making this function public it should be an extension.

Comment (sowmyav27) [reply tests/v2/validation/provisioning/rke2/etcd_backup_restore.go:35]: done

File: tests/framework/extensions/etcdsnapshot/etcdsnapshot.go [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+	_, err = client.Steve.SteveType(clusters.ProvisioningSteveResouceType).Update(existingSteveAPIObj, clusterObj)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getProvisioningClusterByName(client *rancher.Client, clusterName string, namespace string) (*apisV1.Cluster, *steveV1.SteveAPIObject, error) {
```
Comment (anupama2501) [inline tests/framework/extensions/etcdsnapshot/etcdsnapshot.go:45]: Can we remove this. Looks like a spill over from development stage.

**Review by anupama2501 — COMMENTED** [review]
Looks really good. Just 1 comment and can we also get a jenkins run with all the new changes ?

File: tests/framework/extensions/etcdsnapshot/etcdsnapshot.go (hunk 17 above)
Comment (igomez06) [reply tests/framework/extensions/etcdsnapshot/etcdsnapshot.go:45]: agreed

Comment (sowmyav27) [reply tests/framework/extensions/etcdsnapshot/etcdsnapshot.go:45]: oh dang. i missed. ok. thanks

**Review by anupama2501 — DISMISSED** [review]
lgtm

File: tests/framework/extensions/clusters/clusters.go [hunk 18]
```diff
@@ -915,3 +916,18 @@ func WatchAndWaitForCluster(steveClient *v1.Client, kubeProvisioningClient *kube
 	err = wait.WatchWait(result, IsProvisioningClusterReady)
 	return err
 }
+
+func GetProvisioningClusterByName(client *rancher.Client, clusterName string, namespace string) (*apisV1.Cluster, *steveV1.SteveAPIObject, error) {
```
Comment (igomez06) [inline tests/framework/extensions/clusters/clusters.go:920]: Sorry one last thing that I should have mentioned, but won't require another Jenkins run. For our generated docs we write comments for all our exported functions. In this format 
`// <Function> explanation of functions ...`
https://go.dev/doc/comment

**Review by igomez06 — COMMENTED** [review]
Looks great, just need some comments for the new extensions and it's ready.

File: tests/framework/extensions/etcdsnapshot/etcdsnapshot.go [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	steveV1 "github.com/rancher/rancher/tests/framework/clients/rancher/v1"
+	"github.com/rancher/rancher/tests/framework/extensions/clusters"
+)
+
+const (
+	ProvisioningSteveResouceType = "provisioning.cattle.io.cluster"
+)
+
+func CreateSnapshot(client *rancher.Client, clustername string, namespace string) error {
```
[the remark this answers is not in the document]
Comment (sowmyav27) [reply tests/framework/extensions/etcdsnapshot/etcdsnapshot.go:16]: Done

File: tests/framework/extensions/clusters/clusters.go (hunk 18 above)
Comment (sowmyav27) [reply tests/framework/extensions/clusters/clusters.go:921]: Done

**Review by igomez06 — DISMISSED** [review]
LGTM

**Review by anupama2501 — APPROVED** [review]
looks good

File: tests/v2/validation/rbac/rbac_etcd_backup_test.go (hunk 13 above)
Comment (sowmyav27) [reply tests/v2/validation/rbac/rbac_etcd_backup_test.go]: done
